### DISCUSSION TITLE
New Features Wishlist

### DISCUSSION THREAD

**Discussion by wawator** [nebhead/PiTender#14] [Ideas]
Hi,
Thanks again for you help and excellent code!
Here is a few ideas to add to the code
1- have an ajustable flow rate for the fizzy drinks like coke and lemonade. For these drinks I use an air pump to push out the drink out of the bottle, works great and you don't loose the bubbles.
2 - add on the cocktail carrousel the list of ingredients in the drink
What do you think?

**Comment by wawator** [comment]
3 - have the possibility to use the raspberry pi as a hotspot. For example you don't have a WiFi connection. The tablet of phone could connect directly to the WiFi of the raspberry pi. That way you could make cocktails anytime and anywhere you want 😁

**Reply by Jannikovic145** [reply]
I am using my zero W in this way. When I am home (near my home-WIFI), it automatically connects with my home-WIFI and I use it to debug etc. At first I worked with a mobile phone hotspot when I am out, but now my Pi sets up an own WIFI-AP if my Home-WIFI is not available. So I can connect to my Pitender with every mobile device. I am not really sure which manual I used, but try to google something like "auto hotspot switch", there are tons of manuals for different cases :)

Concerning your air pumps: This is exactly what I am searching to use fizzy drinks! Which pumps did you use for this purpose?

**Reply by wawator** [reply]
OK thank I will check the auto switcher system.
For the air pump this what I used 
[Link](https://a.aliexpress.com/_vXN4Rj)

**Reply by wawator** [reply]
Did you use RaspAp?

**Reply by Jannikovic145** [reply]
Thank you, I will check these pumps, this is what my PiTender needs. No, it was defintely not RaspAP. I will check the code for the auto switcher system during the next startup.

**Reply by nebhead** [reply]
Wanted to check _how_ the air pump is used for dispensing carbonated drinks?  Do you put the tubes into the beverage and pump the same way?

**Reply by wawator** [reply]
Hi, this is how I got it to work.
[YouTube link](https://youtu.be/k0ERi0Bt8Gk)
Could do with finding some sort of valve because even after the pump has stopped coke keeps coming out. If you have any ideas 😉

**Reply by nebhead** [reply]
Very interesting!  Yeah, would be good to have a valve that is closed when the relay is off, and then open when the relay / pump are engaged.

**Comment by wawator** [comment]
I have put leds on my machine, is there a way we could implement ws2812b. So that when it's serving a dring it turns red and once its ready it turns green?
And it could have a rainbow animation when nothing is happening. 
This is what I have [Link LED strip](https://a.aliexpress.com/_vEuZSl)

**Comment by nebhead** [comment]
This is great!  I have a few features that I wanted to also add to the list of ideas.  So, I've changed the name of this discussion group to New Features Wishlist.  Anyway, I think we can implement all of these in the future!  I don't have these particular LEDs, but I could look into picking them up and coding something up for them.  

As for the adjustable flow rate for each pump, this is definitely something I want to implement, but hadn't gotten around to it yet.  This will probably be my next addition to the project. 

Adding the ingredients to the carousel is a great idea and I think I can also add this in the future.  

No promises on when I can get these implemented, but will keep them here for my list of improvements.  Thanks for the good suggestions!

**Comment by ardycloud** [comment]
Hi nebhead,

I have tested Pitender in the last few days.
I really like the functions of the WebGui but i miss a Button for a Cup Size and the strength of the cocktail so that the recipe does not have to be adjusted every time.

**Reply by nebhead** [reply]
That's an outstanding idea!  Let's add that to the feature list as well.  Maybe ml volume options for each recipe.

**Reply by ardycloud** [reply]
Also I think about how to possibly do the pre-filling of the hose.
Yes you can "clean" any pump until something comes and then break off. But there must surely be another way.
Also I have now installed a rocker switch to change the polarity to feed the ingredients back into the bottle.
So that nothing is wasted.
Switch is connected exactly if someone also wants to implement it. https://i.ytimg.com/vi/ONUpRZhmdao/maxresdefault.jpg

**Comment by wawator** [comment]
Was thinking it would be nice to change le look of the home screen, was thinking a card slider style. 
Maybe something like this : [Card slider](https://codepen.io/JavaScriptJunkie/pen/WgRBxw)
What do you think?

**Reply by wawator** [reply]
Could even use this as the loading animation [Link](https://codepen.io/MarcMalignan/pen/svLux)

**Reply by nebhead** [reply]
Wow!  This loading animation is perfect!  I wonder if the author would be OK with us using it in this project.

**Reply by wawator** [reply]
Yes it would look great, want me to ask on the codepen page?

**Reply by nebhead** [reply]
Sure!  If they say yes, then I'll get to work updating the project.

**Reply by wawator** [reply]
All good for adding the animation, if you can just add it in the credits?
Thanks its going to look great!

**Reply by nebhead** [reply]
Awesome!  I will see if I can knock this out over the weekend.

**Reply by nebhead** [reply]
Thanks for reaching out to the owner.

**Reply by nebhead** [reply]
And it's done. :)  

https://github.com/nebhead/PiTender/commit/4d4cb0bd05036f8f49a029ff5e0701c4d766221e 

Requires an update to app.py in addition to changing the work.html template, and adding dispense.css.

**Reply by wawator** [reply]
Great work can't wait to see it up and running

**Reply by wawator** [reply]
Just got it set up looks great!

**Comment by wawator** [comment]
Hi, I was wondering if it would be possible to integrate a switch, so that you can only order if a glass is in place. The system could even be used to integrate a [Coin machine ](https://a.aliexpress.com/_uJg2Pv) .

**Reply by nebhead** [reply]
Yep, could definitely implement something like that.  Let's add it to the to-do list. :)

**Reply by wawator** [reply]
Hi there how are you doing?
Any ideas how to add this feature into the code? 🤔

**Reply by nebhead** [reply]
My thought is that a "gate" GPIO pin could be implemented that is sampled before dispensing a drink.  This could be added to the control.py script and could be configured in the app.py settings route.  I don't currently have time to work on this (probably not for the next couple weeks), but if you wanted to give it a shot and do a pull request I'd certainly be happy to incorporate the change.

**Comment by bugoutbilly** [comment]
Someone mentioned using a valve for the carbonated drinks, I found this but it's kinda expensive
https://www.amazon.com/Cole-Parmer-Two-way-normally-closed-solenoid/dp/B003NV1HGY/ref=sr_1_7?crid=EEZR9VSSPQ8F&keywords=pinch+valve&qid=1647383533&sprefix=pinch+valve%2Caps%2C105&sr=8-7

the idea of a pump came across my mind too, but I was afraid of a siphoning effect happening. Also to use air for alcohol would be bad, it tends to advance the decay of it.

I wrote my own code using I2C and this relay, what I like is that its expandable, and the code was fairly simple. So if you wanted to build a machine with 16 pumps or more it would be fairly easy.
https://www.amazon.com/dp/B07KRKS67G?psc=1&ref=ppx_yo2_dt_b_product_details

someone also mentioned a cup size variable, also I instruct people to half fill their cups with ice before selecting their drink. Maybe base a pour off of that AND the flow rate of your pump. a test would be; run each pump for say 10 seconds, times the number of pumps into something that measures ml, so for 8 pumps, 10 seconds, comes out to 80 ml, then you can say the pump does 10ml/sec.

I was tinkering around with adding a cup sensor, by either using a proximity sensor OR a small limit switch. so far no dice

**Comment by Zhangyuan363** [comment]
您好，已收到您的邮件，会尽快给您回复，谢谢。
